webpack: Restart webpack-dev-server on config file changes.

priyank-p opened the pull request:
Using `pyinotify` library automatically restarts webpack process if webpack config files are modified.
Also we are still using python 2 mypy annotations for `tools/webpack`, is this for a reason for this or it just not migrated.

**Testing Plan:** <!-- How have you tested? -->
Run `tools/run-dev.py` then edit any of the webpack config file (`webpack.config.ts`, `webpack-helpers.ts`, `webpack.assets.json`); webpack process should restart.

andersk commented on tools/webpack:
```diff
@@ -55,7 +57,22 @@ def build_for_dev_server(host, port, minify, disable_host_check):
+            # type: (Any) -> None
+            nonlocal webpack_process
+            print('Restarting webpack-dev-sever due to config changes...')
+            webpack_process.kill()
```
`webpack_process.terminate()` would be friendlier. And `webpack_process.wait()` is needed to avoid accumulating zombie processes.

priyank-p replied:
Ah, I tried using `webpack_process.terminate()` first but it did work because I was not using `webpack_process.wait()` and so caused the new webpack process to start before the old one was terminated.

andersk commented on tools/webpack:
```diff
@@ -55,7 +57,22 @@ def build_for_dev_server(host, port, minify, disable_host_check):
+    for file in ['webpack.config.ts', 'webpack-helpers.ts', 'webpack.assets.json']:
+        filepath = os.path.join(os.path.dirname(__file__), file)
+        watch_manager.add_watch(filepath, pyinotify.IN_MODIFY)
+    event_notifier.loop()
```
This should all go in a `try … finally` block that ensures we `webpack_process.terminate(); webpack_process.wait()` at the end.

priyank-p replied:
Just to make sure, this is for making sure webpack process exits cleanly at the end? )(Also curious to why we will need to do this when we weren't doing this before)

andersk replied:
Right. It wasn’t needed before because `os.execvp` replaces the current process with the `webpack` process and does not return to the current script.

andersk commented on tools/webpack:
```diff
@@ -55,7 +57,27 @@ def build_for_dev_server(host, port, minify, disable_host_check):
+
+    class WebpackConfigFileChangeHandler(pyinotify.ProcessEvent):
+        def process_default(self, event):
+            # type: (Any) -> None
```
Avoid `Any` by typing this as `(pyinotify.Event) -> None`.

timabbott commented:
The annotations are just not migrated, feel free to fix that.

timabbott commented:
Merged the first commit; @andersk let me know when you're happy with the rest of this.

timabbott commented:
(I also read the webpack issue tracker to see if they've fixed this, and the answer is definitely no)

While I'm thinking about this, I think probably a good follow-up issue is to see if we can replace `tools/compile-handlebars-templates` with some standard webpack configuration for handlebars; would remove a bit of complexity.

andersk commented on tools/webpack:
```diff
@@ -55,7 +57,27 @@ def build_for_dev_server(host, port, minify, disable_host_check):
     if disable_host_check:
         webpack_args.append('--disable-host-check')
-    os.execvp(webpack_args[0], webpack_args)
+    webpack_process = subprocess.Popen(webpack_args)
```
I’d move this into the `try` block to close a silly race at startup. (Yes that also opens a different silly race, but the different silly race has a more obvious failure mode.)

priyank-p replied:
Can you explain why there would be race conditions?

andersk replied:
In the old version, if we started `webpack` outside the `try` block, and we got a `KeyboardInterrupt` between starting `webpack` and entering the `try` block, we wouldn’t stop `webpack`. (This probably wasn’t likely to happen in practice, but if we can protect against it by reordering some statements, why not.)

andersk commented on tools/webpack:
```diff
@@ -55,7 +57,27 @@ def build_for_dev_server(host, port, minify, disable_host_check):
+        def process_default(self, event):
+            # type: (pyinotify.Event) -> None
+            nonlocal webpack_process
+            print('Restarting webpack-dev-sever due to config changes...')
```
sever → server.

priyank-p commented:
Updated this PR, also added a new commit for using python 3 mypy annotations.

priyank-p commented:
Had to drop the mypy annotation commit because `from typing import NoReturn` was throwing error.

andersk commented:
Looks fine. (It wasn’t necessary to move `class WebpackConfigFileChangeHandler` into the `try` block, but it doesn’t hurt anything either.)

timabbott commented:
Awesome, merged as 2a864ebad3eccf5b02e904c1419a467fab316e9b, after shrinking the `try/except` block as Anders suggested, which I thought made it more readable.  

Thanks @priyank-p and @andersk!  This should be a big quality-of-life improvement.

priyank-p commented:
Hmm, so I moved the class into try block because I though I was getting a linter error (I could be wrong here). Also, I think while merging this you moved the `webpack_process = ..` line back out of the `try` block which @andersk suggested I moved [here](https://github.com/zulip/zulip/pull/12432#discussion_r288857292)

timabbott commented:
I guess I was distracted by the other silly race :).  I'm fine with this working either way, but I think we can't move the class without moving this line, since it accesses the `webpack_process` variable.

andersk commented:
Python has no problem with a class outside the `try` block accessing a variable inside the `try` block. However, MyPy doesn’t understand it. So we should put this back as in dcbc8688ff3f992837e4b1895a125fdb62eeeb0f.

timabbott commented:
Feel free to push a commit updating that.
